sandbox/apparmor/notify/listener/listener.go

```diff
@@ -0,0 +1,365 @@
+// Package listener implements a high-level interface to the apparmor
+// notification mechanism. It can be used to build userspace applications
+// which respond to apparmor prompting profiles.
+package listener
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"sync/atomic"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil/epoll"
+	"github.com/snapcore/snapd/sandbox/apparmor/notify"
+)
+
+var (
+	// ErrListenerAlreadyRun indicates that the Run() method has already been
+	// run. Each listener must only be run once, so that spawned goroutines can
+	// be safely tracked and terminated.
+	ErrListenerAlreadyRun = errors.New("listener has already been run")
+
+	// ErrListenerAlreadyClosed indicates that the listener has been closed
+	ErrListenerAlreadyClosed = errors.New("listener has already been closed")
+
+	// ErrListenerNotSupported indicates that the kernel does not support apparmor prompting
+	ErrListenerNotSupported = errors.New("kernel does not support apparmor notifications")
+
+	osOpen      = os.Open
+	notifyIoctl = notify.Ioctl
+)
+
+// Request is a high-level representation of an apparmor prompting message.
+//
+// Each request must be replied to by writing a boolean to the YesNo channel.
+type Request struct {
+	// Pid is the identifier of the process triggering the request.
+	Pid uint32
+	// Label is the apparmor label on the process triggering the request.
+	Label string
+	// SubjectUID is the UID of the subject that triggered the prompt
+	SubjectUid uint32
+
+	// Path is the path of the file, as seen by the process triggering the request.
+	Path string
+	// Permission is the opaque permission that is being requested.
+	Permission interface{}
+	// YesNo is a channel for writing the response.
+	YesNo chan bool
```

Comment:
this probably should be encapsulated so that the we can expose only a directional channel via a method or something or even has Yes/No methods?

Reply:
I would like to add something like the following:
```go
func (r *Request) Reply(response interface{}) {
        r.replyChan <- response
}
```
and then the listener can cast that response to an appropriate type based on the type of the wait function -- we have different wait functions for different mclasses.

I'll put my ideas in the next commit. Thanks!